qiskit/result/counts.py

```diff
@@ -0,0 +1,29 @@
+# -*- coding: utf-8 -*-
+
+# This code is part of Qiskit.
+#
+# (C) Copyright IBM 2020.
+#
+# This code is licensed under the Apache License, Version 2.0. You may
+# obtain a copy of this license in the LICENSE.txt file in the root directory
+# of this source tree or at http://www.apache.org/licenses/LICENSE-2.0.
+#
+# Any modifications or derivative works of this code must retain this
+# copyright notice, and modified files need to carry a notice indicating
+# that they have been altered from the originals.
+
+from qiskit.result import postprocess
+
+
+class Counts(dict):
```

Comment:
this class is nice.

`qiskit.quantum_info.Statevector.sample_counts()` should also be changed to return this.

Can I also request that two method be added:

`Counts.most_frequent()`: returns the key which has the highest value. This is handy for a lot of circuits where there is one right answer. If there's a tie I'd just raise an error.

`Counts.int_outcomes()`: return another Counts object but with keys converted from bitstrings to integers. Should use little-endian convention for this.

Reply:
Sure, I'll add this on in the next revision I push. The nice thing with this class being a dict subclass `plot_histogram` should just work without any changes because it can be treated as a dict with the same data as the dictionary returned by `get_counts()` as today.

I think I'll actually split this class out into a separate PR now (instead of later like I put in the PR summary) after adding those functions. I was keeping it together for local testing, but with those functions it probably is too much extra scope to include here.

Reply:
Yes, this is very nice! I actually implemented the same thing with the metadata being populated by stuff in the result header and individual experiment data. Works great.

Reply:
The `result.marginalize_counts()` would also be nicely migrated to this as `counts.marginalize()`.

Reply:
I started a separate PR for this here:  #4501